compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETemporaryAllocBufferization.cpp

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-amdaie-temporary-alloc-bufferization"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+static std::optional<BufferOp> createBufferForTemporaryAllocOp(
+    IRRewriter &rewriter, WorkgroupOp workgroupOp, memref::AllocOp allocOp,
+    CoreOp coreOp, unsigned index) {
+  OpBuilder::InsertionGuard g(rewriter);
+  TileOp tileOp = coreOp.getTileOp();
+  // Reset rewriter's location to after last tile's declaration.
+  auto tiles = workgroupOp.getBody()->getOps<TileOp>();
+  assert(!tiles.empty() && "no tiles in workgroupOp");
+  rewriter.setInsertionPointAfter(*std::prev(tiles.end(), 1));
+  auto bufferType = cast<MemRefType>(allocOp.getType());
+  auto bufferOp = rewriter.create<AMDAIE::BufferOp>(
+      rewriter.getUnknownLoc(), bufferType, tileOp, nullptr);
+  return bufferOp;
+}
+
+static LogicalResult bufferizeTemporaryAllocInCoreOp(
+    IRRewriter &rewriter, WorkgroupOp workgroupOp, CoreOp coreOp,
+    SmallVector<Operation *> &toBeErased) {
+  // Step 1. Get all buffers within a CoreOp.
+  // TODO(avarma): All temporary buffers, if present, are at the beginning of
+  // the
+  //               e2e computation within each coreOp and their corresponding
+  //               dealloc ops mark the end of the e2e computation. Therefore we
+  //               are, in order to ensure reuse of the buffer space,
+  //               considering all memref.allocs before the first dealloc op as
+  //               unique set of temporary buffers.
+  SmallVector<memref::AllocOp> allocOps;
+  unsigned numOfUniqueAllocOps = 0;
+  bool foundDeallocOp = false;
+  coreOp.walk([&](Operation *op) {
+    if (auto allocOp = dyn_cast<memref::AllocOp>(op)) {
+      allocOps.push_back(allocOp);
+      toBeErased.push_back(allocOp);
+      if (!foundDeallocOp) numOfUniqueAllocOps++;
+    } else if (auto deallocOp = dyn_cast<memref::DeallocOp>(op)) {
+      toBeErased.push_back(deallocOp);
+      foundDeallocOp = true;
```

Comment:
Can't you just look at the users of the `AllocOp` to find the `Dealloc`? And shouldn't we just expect every alloc to have a dealloc? This would simplify the logic below (with `numOfUniqueAllocOps`) as well I think.

Reply:
Every alloc is indeed having a dealloc op.
But, the main intention of `numOfUniqueAllocOps` was to reuse the temporary buffers. Example :

aie.core {
     %a = memref.alloc
     <SOME USE>
     dealloc %a
     %b = memref.alloc
     <SAME USE LIKE ABOVE>
     dealloc %b
}
So, we shouldn't need to re-create `amdaie.buffer` for `%b` . We can reuse the buffer created for `%a`. So, the logic I added with `foundDeallocOp` is just to get hold of the unique set of AllocOps (the test case I added highlights this well).

I don't think simply fetching Dealloc user of Alloc is solving the `numOfUniqueAllocOps`  logic for reuse.

Reply:
> So, we shouldn't need to re-create amdaie.buffer for %b . We can reuse the buffer created for %a. So, the logic I added with foundDeallocOp is just to get hold of the unique set of AllocOps (the test case I added highlights this well).

I see. In that case, you need to check the sizes of the allocs as well and keep track of which alloc the dealloc belongs to and I don't see that logic?

Reply:
That logic is fairly more involved than what we need so I am not sure why we're adding it? And it might already exist on `memref.alloc` somewhere else?

Reply:
As discussed offline, I've removed the optimization on temporary buffer that was being done with certain assumptions.